### PR TITLE
Implement MapWindow::assignRouteColours

### PR DESCRIPTION
**Pull request by AaronVanGeffen**
WIP

### CODE REVIEW

File: src/OpenLoco/src/Windows/MapWindow.cpp [hunk 1]
```diff
@@ -1508,11 +1510,92 @@ namespace OpenLoco::Ui::Windows::MapWindow
         }
     }
 
-    // 0x0046CED0
-    static void sub_46CED0()
+    // 0x00478265
+    static void sub_478265(int8_t* buffer)
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/Windows/MapWindow.cpp:1514]: I'm not really a fan of marking this as signed. Its just that road elements have the highest bit set (and the end of the array is 0xFF). Its not really signed.

Comment (duncanspumpkin) [reply src/OpenLoco/src/Windows/MapWindow.cpp:1514]: In addition the other user of sub_478265 using uint8_t

File: src/OpenLoco/src/Windows/MapWindow.cpp [hunk 2]
```diff
@@ -1508,11 +1510,92 @@ namespace OpenLoco::Ui::Windows::MapWindow
...
+        auto availableTracks = CompanyManager::getPlayerCompany()->getAvailableRailTracks();
+
+        std::array<int8_t, 100> buffer = {};
+        assert(std::size(buffer) >= std::size(availableTracks));
+
+        std::copy(std::begin(availableTracks), std::end(availableTracks), std::begin(buffer));
+        buffer[availableTracks.size()] = std::numeric_limits<uint8_t>::max();
+
+        auto* nextEl = &buffer[availableTracks.size()];
+        sub_478265(nextEl);
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/Windows/MapWindow.cpp:1567]: Technically this function should also have updatingCompanyId set to player company for the duration as well but i really don't think that is needed.

Comment (AaronVanGeffen) [reply src/OpenLoco/src/Windows/MapWindow.cpp:1567]: Yeah, I took that as a means to accommodate the function that is no longer necessary in our codebase.

File: src/OpenLoco/src/Windows/MapWindow.cpp [hunk 3]
```diff
@@ -1508,11 +1510,92 @@ namespace OpenLoco::Ui::Windows::MapWindow
...
+                _byte_F253DF[i] = buffer[i];
+                auto freeColour = std::max(0, Numerics::bitScanForward(availableColours));
+                availableColours &= ~(1U << freeColour);
+
+                auto colour = industryColours[freeColour];
+                _routeColours[i] = colour;
+
+                if (buffer[i] & 0x80)
+                {
+                    _byte_F2540C[i & 0x7F] = colour;
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/Windows/MapWindow.cpp:1582]: buffer[i] & 0x7F

File: src/OpenLoco/src/Windows/MapWindow.cpp [hunk 4]
```diff
@@ -1508,11 +1510,92 @@ namespace OpenLoco::Ui::Windows::MapWindow
...
+                }
+                else
+                {
+                    _byte_F25404[i] = colour;
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/Windows/MapWindow.cpp:1586]: buffer[i]

**Review by LeftofZen — APPROVED** [review]
map colours for company track+vehicles appear correct

File: src/OpenLoco/src/Windows/MapWindow.cpp (hunk 2 above)
Comment (ZehMatt) [reply src/OpenLoco/src/Windows/MapWindow.cpp:1567]: sub_478265 checks for unlocked vehicles using updatingCompanyId, is it assigned to the player at this point or why don't we need it?

Comment (ZehMatt) [reply src/OpenLoco/src/Windows/MapWindow.cpp:1567]: That would be probably best or we assert that the updating company id is that of a player.

Comment (AaronVanGeffen) [reply src/OpenLoco/src/Windows/MapWindow.cpp:1567]: Following up in #2094

**Pull request merged**